sys/net/link_layer/ieee802154/submac.c

```diff
@@ -0,0 +1,384 @@
+    submac->panid = CONFIG_IEEE802154_DEFAULT_PANID;
+
+    submac->be.min = CONFIG_IEEE802154_DEFAULT_CSMA_CA_MIN_BE;
+    submac->csma_retries = CONFIG_IEEE802154_DEFAULT_CSMA_CA_RETRIES;
+    submac->be.max = CONFIG_IEEE802154_DEFAULT_CSMA_CA_MAX_BE;
+
+    submac->tx_pow = CONFIG_IEEE802154_DEFAULT_TXPOWER;
+
+    if (ieee802154_radio_has_24_ghz(dev)) {
+        submac->channel_num = CONFIG_IEEE802154_DEFAULT_CHANNEL;
+        submac->channel_page = 0;
+    }
+    else {
+        submac->channel_num = CONFIG_IEEE802154_DEFAULT_SUBGHZ_CHANNEL;
+        submac->channel_page = CONFIG_IEEE802154_DEFAULT_SUBGHZ_PAGE;
+    }
+
+    /* If the radio is still not in TRX_OFF state, spin */
+    while (ieee802154_radio_confirm_on(dev) == -EAGAIN) {}
+
+    /* Enable Auto ACK */
+    ieee802154_radio_set_rx_mode(dev, IEEE802154_RX_AACK_ENABLED);
+
+    /* Configure address filter */
+    ieee802154_radio_set_hw_addr_filter(dev, &submac->short_addr,
+                                        &submac->ext_addr, &submac->panid);
+
+    /* Configure PHY settings (channel, TX power) */
+    ieee802154_phy_conf_t conf =
+    { .channel = CONFIG_IEEE802154_DEFAULT_CHANNEL,
+      .page = CONFIG_IEEE802154_DEFAULT_CHANNEL,
+      .pow = CONFIG_IEEE802154_DEFAULT_TXPOWER };
+
+    ieee802154_radio_config_phy(dev, &conf);
+    assert(ieee802154_radio_set_cca_threshold(dev,
+                                              IEEE802154_CCA_THRESH_DEFAULT) >=
+           0);
+
+    ieee802154_radio_request_set_trx_state(dev, IEEE802154_TRX_STATE_RX_ON);
+
+    return 0;
+}
+
+int ieee802154_set_phy_conf(ieee802154_submac_t *submac, uint16_t channel_num,
+                            uint8_t channel_page, int8_t tx_pow)
+{
+    ieee802154_dev_t *dev = submac->dev;
+    ieee802154_phy_conf_t conf =
+    { .channel = channel_num, .page = channel_page, .pow = tx_pow };
+
+    if (submac->state == IEEE802154_STATE_OFF) {
+        return -ENETDOWN;
+    }
+
+    int res = ieee802154_radio_config_phy(dev, &conf);
+
+    if (res >= 0) {
+        submac->channel_num = channel_num;
+        submac->channel_page = channel_page;
+        submac->tx_pow = tx_pow;
+    }
+
+    return res;
+}
+
+int ieee802154_set_state(ieee802154_submac_t *submac, ieee802154_submac_state_t state)
+{
+    int res;
+
+    ieee802154_dev_t *dev = submac->dev;
+
+    if (submac->tx) {
+        return -EBUSY;
+    }
+
+    if (state == submac->state) {
+        return -EALREADY;
+    }
+
+    /* Wake up the radio if it was off */
+    if (submac->state == IEEE802154_STATE_OFF) {
+        if ((res = ieee802154_radio_request_on(dev)) < 0) {
+            return res;
+        }
+        while (ieee802154_radio_confirm_on(dev) == -EAGAIN);
+    }
+
+    if (state == IEEE802154_STATE_OFF) {
+        res = ieee802154_radio_off(dev);
+    }
+    else {
+        if ((res = ieee802154_radio_request_set_trx_state(dev, state == IEEE802154_STATE_IDLE ? IEEE802154_TRX_STATE_TRX_OFF : IEEE802154_TRX_STATE_RX_ON)) < 0) {
```

Comment:
Maybe something like

```suggestion

        ieee802154_submac_state_t new_state =
                   state == IEEE802154_STATE_IDLE
                   ? IEEE802154_TRX_STATE_TRX_OFF
                   : IEEE802154_TRX_STATE_RX_ON;
        if ((res = ieee802154_radio_request_set_trx_state(dev, new_state) < 0) {
```